File: src/lexer.js
const { keywords, operators } = require('./token')

const SYMBOLS = '.%^&|:=~+-*<>!/'

const lexer = (source) => ({
  source,
  line: 1,
  curr: 0,
  prev: 0,
  mode: 'regular',
  tokens: [],
})

const token = (lexer, kind, value = null) => {
  lexer.tokens.push({
    kind,
    value,
    span: {
      lineno: lexer.line,
      start: lexer.prev,
      offset: lexer.curr,
    }
  })
  return lexer
}

const done = (lexer) => {
  return lexer.curr >= lexer.source.length
}

const bump = (lexer) => {
  lexer.curr += 1
  return lexer
}

const line = (lexer) => {
  lexer.line += 1
  return bump(lexer)
}

const current = (lexer) => {
  return lexer.source[lexer.curr]
}

const peek = (lexer) => {
  return lexer.source[lexer.curr + 1]
}

const slice = (lexer) => {
  return lexer.source.slice(lexer.prev, lexer.curr)
}

const numeric = (char) => {
  return char >= '0' && char <= '9'
}

const lower = (char) => {
  return char >= 'a' && char <= 'z' || char === '_'
}

const upper = (char) => {
  return char >= 'A' && char <= 'Z'
}

const alpha = (char) => {
  return lower(char) || upper(char)
}

const alphanumeric = (char) => {
  return lower(char) || upper(char) || numeric(char)
}

const symbolic = (char) => {
  return SYMBOLS.includes(char)
}

const error = (lexer, msg) => {
  throw `Error [${lexer.line}]: ${msg}`
}

const comment = (lexer) => {
  while (current(lexer) !== '\n' && !done(lexer)) {
    bump(lexer)
  }
  return lexer
}

const number = (lexer, prefix = false) => {
  if (prefix) {
    bump(lexer)
  }

  let float = false
  while (numeric(current(lexer))) {
    bump(lexer)
    if (current(lexer) === '.' && !float) {
      float = true
      bump(lexer)
    }
  }
  return token(lexer, 'number', slice(lexer))
}

const identifier = (lexer) => {
  while (alphanumeric(current(lexer))) {
    bump(lexer)
  }

  switch (current(lexer)) {
    case '?':
    case '!': {
      bump(lexer)
    }
  }

  const ident = slice(lexer)

  const keyword = keywords[ident]

  if (keyword) {
    return token(lexer, keyword)
  } else {
    return token(lexer, upper(ident[0]) ? 'upper' : 'lower', ident)
  }
}

const special = (lexer) => {
  bump(lexer)

  while (alphanumeric(current(lexer))) {
    bump(lexer)
  }

  switch (slice(lexer)) {
    case '@js':
      return 'native'
    default:
      return error(lexer, `invalid special form ${slice(lexer)}`)
  }
}

const operator = (lexer) => {
  while (symbolic(current(lexer))) {
    bump(lexer)
  }

  const operator = operators[slice(lexer)]

  if (operator) {
    return token(lexer, operator)
  } else {
    return error(lexer, `Invalid operator '${slice(lexer)}'`)
  }

}

const next = (lexer) => {
  lexer.prev = lexer.curr

  if (lexer.mode === 'string') {
    return template(lexer, false)
  }

  const char = current(lexer)

  switch (char) {
    case ' ':
    case '\t':
    case '\r': return bump(lexer)
    case '\n': return line(lexer)
    case ',':  return token(bump(lexer), 'comma')
    case ';':  return token(bump(lexer), 'semi')
    case '(':  return token(bump(lexer), 'lparen')
    case ')':  return token(bump(lexer), 'rparen')
    case '[':  return token(bump(lexer), 'lbracket')
    case ']':  return token(bump(lexer), 'rbracket')
    case '{':  return token(bump(lexer), 'lbrace')
    case '}':  {
      if (lexer.mode === 'template') {
          lexer.mode = 'string'
      }
      return token(bump(lexer), 'rbrace')
    }
    case '"':
      return template(lexer, true)
    case '+':
    case '-':
      if (numeric(peek(lexer))) {
        return number(lexer, true)
      } else {
        return operator(lexer)
      }
    case '/':
      if (peek(lexer) === '/') {
        return comment(lexer)
      } else {
        return operator(lexer)
      }
    case '@':
      return special(lexer)
    default:
      if (alpha(char)) {
        return identifier(lexer)
      } else if (numeric(char)) {
        return number(lexer)
      } else if (symbolic(char)) {
        return operator(lexer)
      } else {
        bump(lexer)
        return error(lexer, `Invalid character '${slice(lexer)}'`)
      }
  }
}

const template = (lexer, start) => {
  lexer.prev = lexer.curr

  while (!done(lexer)) {
    switch (current(lexer)) {
      case '"': {
        if (lexer.mode === 'regular') {
          bump(lexer)
          lexer.mode = 'string'
        } else {
          lexer.mode = 'regular'
          bump(lexer)
          const kind = start ? 'string' : 'string finish'
          return token(lexer, kind, slice(lexer))
        }
      }
        break
      case '{': {
        if (peek(lexer) === '{') {
          bump(lexer)
          bump(lexer)
        } else {
          const kind = start ? 'string start' : 'string fragment'
          lexer.mode = 'template'
          return token(lexer, kind, slice(lexer))
        }
      }
        break
      case '\n':
        return error(lexer, 'Unterminated string')
      case '\\': {
        bump(lexer)
        switch (current(lexer)) {
          case 'n': case 'r':
          case 't': case 'v':
          case 'a': case 'b':
          case '"': case '0':
          case '\\': {
            bump(lexer)
          }
            break
          case 'u':
            return error(lexer, 'TODO Validate unicode escape')
          case 'x':
            return error(lexer, 'TODO Validate binary escape')
          default:
            return error(lexer, `Invalid escape character ${current(lexer)}`)
        }
      }
        break
      default:
        bump(lexer)
    }
  }

}

exports.tokenize = (source) => {
  const tokenizer = lexer(source)

  while (!done(tokenizer)) { next(tokenizer) }

  return tokenizer.tokens
}

File: src/lexer.rs
use crate::source::{Source, Span};

use super::token::{Token, TokenKind};

#[derive(Debug, Clone)]
pub struct Lexer<'l> {
    source: &'l Source,
    bytes: &'l [u8],
    curr: u8,
    peek: u8,
    index: usize,
    start: usize,
}

impl<'l> Lexer<'l> {
    fn new(source: &'l Source) -> Lexer<'l> {
        let mut lexer = Lexer {
            source,
            bytes: source.content.as_bytes(),
            curr: 0,
            peek: 0,
            index: 0,
            start: 0,
        };
        lexer.curr = lexer.at(0);
        lexer.peek = lexer.at(1);
        lexer
    }

    fn align(&mut self) {
        self.start = self.index;
    }

    fn bump(&mut self) -> u8 {
        self.index += 1;
        self.curr = self.peek;
        self.peek = self.at(self.index + 1);
        self.curr
    }

    fn at(&self, idx: usize) -> u8 {
        if idx < self.bytes.len() {
            self.bytes[idx]
        } else {
            b'\0'
        }
    }

    fn slice(&self) -> &str {
        &self.source.content[self.start..self.index]
    }

    fn span(&self) -> Span {
        Span::new(self.start as u32, self.index as u32)
    }

    fn bump_while<P>(&mut self, pred: P)
    where
        P: Fn(u8) -> bool,
    {
        while pred(self.curr) {
            self.bump();
        }
    }

    fn trivia(&mut self) {
        loop {
            if self.curr.is_ascii_whitespace() && self.curr != b'\n' {
                self.bump_while(|chr| chr != b'\n' && chr.is_ascii_whitespace());
            } else if self.curr == b'/' && self.peek == b'/' {
                self.bump_while(|chr| chr != b'\n' && chr != b'\0');
            } else {
                break;
            }
        }
    }
}

pub fn tokens(source: &Source) -> Vec<Token> {
    let mut lexer = Lexer::new(source);
    let mut tokens = vec![];

    while let kind = kind(&mut lexer) && kind != TokenKind::EOF {
        tokens.push(Token {
            kind,
            span: lexer.span(),
        });
    }
    
    tokens
}

fn kind(l: &mut Lexer) -> TokenKind {
    l.trivia();
    l.align();

    let kind = match l.curr {
        b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
            return ident(l);
        }
        b'0'..=b'9' => {
            return number(l);
        }
        b'"' => {
            return string(l);
        }
        b'\n' => {
            return newline(l);
        }
        _ if is_symbol(l.curr) => {
            return operator(l);
        }
        b'(' => TokenKind::LParen,
        b')' => TokenKind::RParen,
        b'{' => TokenKind::LBrace,
        b'}' => TokenKind::RBrace,
        b'[' => TokenKind::LBracket,
        b']' => TokenKind::RBracket,
        b';' => TokenKind::Semi,
        b',' => TokenKind::Comma,
        0 => TokenKind::EOF,
        _ => TokenKind::UnexpectedCharacter,
    };
    l.bump();
    kind
}

fn ident(l: &mut Lexer) -> TokenKind {
    l.bump_while(|chr| matches!(chr, b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_'));
    l.bump_while(|chr| chr == b'\'');

    let value = l.slice();

    match value {
        "and" => TokenKind::And,
        "or" => TokenKind::Or,
        "not" => TokenKind::Not,
        _ => {
            if value == "_" {
                TokenKind::Underscore
            } else {
                TokenKind::Identifier
            }
        }
    }
}

fn is_symbol(chr: u8) -> bool {
    matches!(
        chr,
        b'=' | b':' | b'.' | b'|' | b'<' | b'>' | b'!' | b'+' | b'-' | b'*' | b'/' | b'^' | b'%'
    )
}

fn operator(l: &mut Lexer) -> TokenKind {
    let kind = match l.curr {
        b'.' if l.peek == b'.' => {
            l.bump();
            TokenKind::Dots
        }
        b'.' => TokenKind::Dot,
        b'=' if l.peek == b'=' => {
            l.bump();
            TokenKind::EqEq
        }
        b'=' if l.peek == b'>' => {
            l.bump();
            TokenKind::Arrow
        }
        b'=' => TokenKind::Equals,
        b':' => TokenKind::Colon,
        b'<' if l.peek == b'=' => {
            l.bump();
            TokenKind::LtEq
        }
        b'<' => TokenKind::Lt,
        b'>' if l.peek == b'=' => {
            l.bump();
            TokenKind::GtEq
        }
        b'>' => TokenKind::Gt,
        b'!' if l.peek == b'=' => {
            l.bump();
            TokenKind::NoEq
        }
        b'+' => TokenKind::Plus,
        b'-' => TokenKind::Minus,
        b'*' => TokenKind::Star,
        b'/' => TokenKind::Slash,
        b'^' => TokenKind::Caret,
        b'%' => TokenKind::Mod,
        _ => unreachable!(),
    };

    l.bump();
    kind
}

fn number(l: &mut Lexer) -> TokenKind {
    l.bump_while(|chr| chr.is_ascii_digit());

    match l.curr {
        b'.' if l.peek.is_ascii_digit() => {
            l.bump();
            l.bump_while(|chr| chr.is_ascii_digit());
        }
        b'b' if l.slice() == "0" => {
            l.bump();
            l.bump_while(|chr| matches!(chr, b'0' | b'1'));
        }
        b'x' if l.slice() == "0" => {
            l.bump();
            l.bump_while(|chr| chr.is_ascii_hexdigit());
        }
        _ => return TokenKind::Integer,
    }

    TokenKind::Number
}

fn string(l: &mut Lexer) -> TokenKind {
    loop {
        match l.bump() {
            b'\n' | 0 => {
                return TokenKind::UnterminatedString;
            }
            b'"' => {
                l.bump();
                break;
            }
            b'\\' => match l.bump() {
                b'\\' | b'n' | b'r' | b't' | b'"' | b'0' => {}
                b'u' => todo!("validate escaped unicode"),
                b'x' => todo!("validate escaped binary"),
                _ => {
                    return TokenKind::UnexpectedCharacter;
                }
            },
            _ => {}
        }
    }

    TokenKind::String
}

fn newline(l: &mut Lexer) -> TokenKind {
    while l.curr == b'\n' {
        l.bump();
        l.trivia();
    }
    TokenKind::NewLine
}
